components/role-mgt/org.wso2.carbon.identity.role.mgt.core/src/main/java/org/wso2/carbon/identity/role/mgt/core/RoleManagementServiceImpl.java

```diff
@@ -0,0 +1,208 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.role.mgt.core;
+
+import org.apache.commons.lang.ArrayUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.context.CarbonContext;
+import org.wso2.carbon.identity.core.util.IdentityTenantUtil;
+import org.wso2.carbon.identity.role.mgt.core.dao.RoleDAO;
+import org.wso2.carbon.identity.role.mgt.core.dao.RoleDAOImpl;
+import org.wso2.carbon.user.core.UserRealm;
+import org.wso2.carbon.user.mgt.UserRealmProxy;
+import org.wso2.carbon.user.mgt.common.UIPermissionNode;
+import org.wso2.carbon.user.mgt.common.UserAdminException;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.wso2.carbon.identity.role.mgt.core.RoleConstants.Error.UNEXPECTED_SERVER_ERROR;
+
+/**
+ * Implementation of the {@link RoleManagementService} interface.
+ */
+public class RoleManagementServiceImpl implements RoleManagementService {
+
+    private static final Log log = LogFactory.getLog(RoleManagementServiceImpl.class);
+
+    @Override
+    public RoleBasicInfo addRole(String roleName, List<String> userList, List<String> groupList,
+            List<String> permissions, String tenantDomain) throws IdentityRoleManagementException {
+
+        RoleDAO roleDAO = RoleDAOImpl.getInstance();
+        RoleBasicInfo role = roleDAO.addRole(roleName, userList, groupList, tenantDomain);
+        try {
+            getUserAdminProxy().setRoleUIPermission(roleName, permissions.toArray(new String[0]));
+        } catch (UserAdminException e) {
+            throw new IdentityRoleManagementServerException(UNEXPECTED_SERVER_ERROR.getCode(),
+                    "An error occurred when setting permissions for the role: " + roleName, e);
+        }
+        return role;
+    }
+
+    @Override
+    public List<RoleBasicInfo> getRoles(int limit, int offset, String sortBy, String sortOrder, String tenantDomain)
+            throws IdentityRoleManagementException {
+
+        RoleDAO roleDAO = RoleDAOImpl.getInstance();
+        return roleDAO.getRoles(limit, offset, sortBy, sortOrder, tenantDomain);
+    }
+
+    @Override
+    public List<RoleBasicInfo> getRoles(String filter, int limit, int offset, String sortBy, String sortOrder,
+            String tenantDomain) throws IdentityRoleManagementException {
+
+        RoleDAO roleDAO = RoleDAOImpl.getInstance();
+        return roleDAO.getRoles(filter, limit, offset, sortBy, sortOrder, tenantDomain);
+    }
+
+    @Override
+    public Role getRole(String roleID, String tenantDomain) throws IdentityRoleManagementException {
+
+        RoleDAO roleDAO = RoleDAOImpl.getInstance();
+        Role role = roleDAO.getRole(roleID, tenantDomain);
+        role.setPermissions(getPermissionListOfRole(roleID, tenantDomain));
```

Comment:
Doesn't this trigger duplicated queries
Can't this be addressed as dao layer

Reply:
We have used the available UserRealmProxy for mange permissions. So we didn't implement a DAO for that in the role management component.

Reply:
This is the existing RolePermissionManagementServiceImpl. We have moved the same implementation to the role management service. https://github.com/wso2/carbon-identity-framework/blob/master/components/user-mgt/org.wso2.carbon.user.mgt/src/main/java/org/wso2/carbon/user/mgt/RolePermissionManagementServiceImpl.java

Reply:
Using user realm proxy is fine, but this initiate same query to convert role name right

Reply:
Ah yes. Got the point. Moved the setPermissions logic to DAO and updated the PR.